### ISSUE TITLE
BlackHole and ffmpeg

### ISSUE THREAD

**Issue by fuerst** [ExistentialAudio/BlackHole#45]
Recording audio with `ffmpeg` through BlackHole works quite well, thanks for this!

I encounter one problem though: The second execution of `ffmpeg` does not work. There is 3 CTRL-C necessary to kill `ffmpeg`. It then responds with: _Received > 3 system signals, hard exiting_. After that, the next `ffmpeg` command does work, the `ffmpeg` after that not and so on. 

Killing `coreaudiod` before the second `ffmpeg` does help too.

It happens only if audio is or was playing at the first `ffmpeg` command.

I tested `ffmpeg` version 4.2.2
You may test it using this command:

```
# Get input channel number for BlackHole device
ffmpeg -f avfoundation -list_devices true -i ""

# Record 2 seconds from i.e. input channel 0
ffmpeg -loglevel debug -f avfoundation -i :0 -t 2 test.mp3
```

`-loglevel debug` shows much, interesting is this:

First ffmpeg call:

```
...
[avfoundation @ 0x7f8843000000] audio device 'BlackHole 16ch' opened
[avfoundation @ 0x7f8843000000] All info found
...
```

Second ffmpeg call stops at _audio device 'BlackHole 16ch' opened_

Recording through other input channels like _Built-in Microphone_ does work multiple times, as expected. 

Any idea what may be wrong here?

**Comment by devinroth** [comment]
Can't think of what is causing the problem immediately. Seems strange. So something is happening at the end of the first record that is causing the problem. I haven't used ffmpeg before. I'll dig into this deeper when I have time.

**Comment by fuerst** [comment]
Thanks for that!

Interesting finding by accident: I opened a _New Audio Recording_ in Quicktime Player and choose BlackHole there as Input (called _Microphone_). Without active recording in Quicktime Player the multiple `ffmpeg` commands did work without problems. So at least there is a workaround.

May be `ffmpeg` does some unexpected things when closing the BlackHole audio device which prevents it from reopening again.

Two more findings:

* the problem occurs only if BlackHole is selected as default output device.
* Quicktime Player shortly display a yellow warning sign near the record button when opening a _New Audio Recording_ after one successful `ffmpeg` run.

**Comment by devinroth** [comment]
Super helpful. Does anything come up on Console? You can search for coreaudio.

**Comment by devinroth** [comment]
And how does it compare to using a different audio input not BlackHole or the build in mic?

**Comment by fuerst** [comment]
Build in mic works every time, no problem.

I see the following errors in the Console when the faulty `ffmpeg` runs. 
Do you want to see the whole log of a working and the next faulty call?

```
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALB_MachPort.cpp:277:SendMessage:  HALB_MachPort::SendMessage: mach_msg() returned error 0x10000003: '(ipc/send) invalid destination port'
ffmpeg: (CMIOBaseUnits) [com.apple.cmio:] CMIO_Unit_Input_HAL.cpp:985:GetPropertyInfo CMIOUInputBase::GetPropertyInfo failed for ID 102, Error: -67456

ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALB_MachPort.cpp:277:SendMessage:  HALB_MachPort::SendMessage: mach_msg() returned error 0x10000003: '(ipc/send) invalid destination port'
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ProxyIOContext.cpp:1399:IOWorkLoop:  HALC_ProxyIOContext::IOWorkLoop: failed to send the final message to the server, Error: 0x10000003
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ShellPlugIn.cpp:817:HAL_HardwarePlugIn_ObjectHasProperty:  HAL_HardwarePlugIn_ObjectHasProperty: no object
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ShellPlugIn.cpp:817:HAL_HardwarePlugIn_ObjectHasProperty:  HAL_HardwarePlugIn_ObjectHasProperty: no object
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ShellPlugIn.cpp:817:HAL_HardwarePlugIn_ObjectHasProperty:  HAL_HardwarePlugIn_ObjectHasProperty: no object
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ShellPlugIn.cpp:817:HAL_HardwarePlugIn_ObjectHasProperty:  HAL_HardwarePlugIn_ObjectHasProperty: no object
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ShellPlugIn.cpp:817:HAL_HardwarePlugIn_ObjectHasProperty:  HAL_HardwarePlugIn_ObjectHasProperty: no object
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ShellPlugIn.cpp:817:HAL_HardwarePlugIn_ObjectHasProperty:  HAL_HardwarePlugIn_ObjectHasProperty: no object

coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_UCPlugIn.cpp:91:Construct_New:  HALS_UCPlugIn::Construct_New: failed to find a bundle ID
coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_UCPlugIn.cpp:169:Construct_Old:  HALS_UCPlugIn::Construct_Old: failed to find a bundle ID
coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_UCPlugIn.cpp:46:HALS_UCPlugIn:  HALS_UCPlugIn::HALS_UCPlugIn: construction failed, probably because this isn't a server plug-in, Error: 1852797029 (nope)

ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ShellPlugIn.cpp:817:HAL_HardwarePlugIn_ObjectHasProperty:  HAL_HardwarePlugIn_ObjectHasProperty: no object
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ShellPlugIn.cpp:817:HAL_HardwarePlugIn_ObjectHasProperty:  HAL_HardwarePlugIn_ObjectHasProperty: no object
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ShellPlugIn.cpp:817:HAL_HardwarePlugIn_ObjectHasProperty:  HAL_HardwarePlugIn_ObjectHasProperty: no object

ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ShellPlugIn.cpp:1118:HAL_HardwarePlugIn_DeviceStop:  HAL_HardwarePlugIn_DeviceStart: no device
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALPlugIn.cpp:536:DeviceStopIOProc:  HALPlugIn::StopIOProc: got an error from the plug-in routine, Error: 560947818 (!obj)
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ShellPlugIn.cpp:996:HAL_HardwarePlugIn_DeviceDestroyIOProcID:  HAL_HardwarePlugIn_DeviceDestroyIOProcID: no device
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALPlugIn.cpp:454:DeviceDestroyIOProcID:  HALPlugIn::DeviceCreateIOProcID: got an error from the plug-in routine, Error: 560947818 (!obj)
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ShellPlugIn.cpp:817:HAL_HardwarePlugIn_ObjectHasProperty:  HAL_HardwarePlugIn_ObjectHasProperty: no object
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ShellPlugIn.cpp:817:HAL_HardwarePlugIn_ObjectHasProperty:  HAL_HardwarePlugIn_ObjectHasProperty: no object
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ShellPlugIn.cpp:817:HAL_HardwarePlugIn_ObjectHasProperty:  HAL_HardwarePlugIn_ObjectHasProperty: no object
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ShellPlugIn.cpp:817:HAL_HardwarePlugIn_ObjectHasProperty:  HAL_HardwarePlugIn_ObjectHasProperty: no object
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ShellPlugIn.cpp:817:HAL_HardwarePlugIn_ObjectHasProperty:  HAL_HardwarePlugIn_ObjectHasProperty: no object
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ShellPlugIn.cpp:817:HAL_HardwarePlugIn_ObjectHasProperty:  HAL_HardwarePlugIn_ObjectHasProperty: no object
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ShellPlugIn.cpp:817:HAL_HardwarePlugIn_ObjectHasProperty:  HAL_HardwarePlugIn_ObjectHasProperty: no object
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ShellPlugIn.cpp:817:HAL_HardwarePlugIn_ObjectHasProperty:  HAL_HardwarePlugIn_ObjectHasProperty: no object
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ShellPlugIn.cpp:817:HAL_HardwarePlugIn_ObjectHasProperty:  HAL_HardwarePlugIn_ObjectHasProperty: no object
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ShellPlugIn.cpp:817:HAL_HardwarePlugIn_ObjectHasProperty:  HAL_HardwarePlugIn_ObjectHasProperty: no object
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ShellPlugIn.cpp:817:HAL_HardwarePlugIn_ObjectHasProperty:  HAL_HardwarePlugIn_ObjectHasProperty: no object
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] AudioHardware.cpp:1210:AudioObjectRemovePropertyListener:  AudioObjectRemovePropertyListener: no object with given ID 59
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] AudioHardware.cpp:1210:AudioObjectRemovePropertyListener:  AudioObjectRemovePropertyListener: no object with given ID 59
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] AudioHardware.cpp:1210:AudioObjectRemovePropertyListener:  AudioObjectRemovePropertyListener: no object with given ID 59
ffmpeg: (CMIOBaseUnits) [com.apple.cmio:] CMIO_Unit_ScopeElement.h:200:SafeGetElement Throwing err: -67454
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] AudioHardware.cpp:1210:AudioObjectRemovePropertyListener:  AudioObjectRemovePropertyListener: no object with given ID 59
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] AudioHardware.cpp:1210:AudioObjectRemovePropertyListener:  AudioObjectRemovePropertyListener: no object with given ID 59
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] AudioHardware.cpp:1210:AudioObjectRemovePropertyListener:  AudioObjectRemovePropertyListener: no object with given ID 59
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ProxySystem.cpp:143:GetObjectInfo:  HALC_ProxySystem::GetObjectInfo: got an error from the server, Error: 560947818 (!obj)
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ProxySystem.cpp:143:GetObjectInfo:  HALC_ProxySystem::GetObjectInfo: got an error from the server, Error: 560947818 (!obj)
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ProxySystem.cpp:143:GetObjectInfo:  HALC_ProxySystem::GetObjectInfo: got an error from the server, Error: 560947818 (!obj)
ffmpeg: (CoreAudio) [com.apple.coreaudio:AudioHAL_Client] HALC_ProxySystem.cpp:143:GetObjectInfo:  HALC_ProxySystem::GetObjectInfo: got an error from the server, Error: 560947818 (!obj)

coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_MIGServer.cpp:1806:HALS_Object_GetPropertyData_DAI32:  HALS_Object_GetPropertyData_DAI32: the object does not implement the property
coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_MIGServer.cpp:1806:HALS_Object_GetPropertyData_DAI32:  HALS_Object_GetPropertyData_DAI32: the object does not implement the property
coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_MIGServer.cpp:1806:HALS_Object_GetPropertyData_DAI32:  HALS_Object_GetPropertyData_DAI32: the object does not implement the property
coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_MIGServer.cpp:1806:HALS_Object_GetPropertyData_DAI32:  HALS_Object_GetPropertyData_DAI32: the object does not implement the property
coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_MIGServer.cpp:1806:HALS_Object_GetPropertyData_DAI32:  HALS_Object_GetPropertyData_DAI32: the object does not implement the property
coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_MIGServer.cpp:1806:HALS_Object_GetPropertyData_DAI32:  HALS_Object_GetPropertyData_DAI32: the object does not implement the property
coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_MIGServer.cpp:1806:HALS_Object_GetPropertyData_DAI32:  HALS_Object_GetPropertyData_DAI32: the object does not implement the property
coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_MIGServer.cpp:1806:HALS_Object_GetPropertyData_DAI32:  HALS_Object_GetPropertyData_DAI32: the object does not implement the property
coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_MIGServer.cpp:1806:HALS_Object_GetPropertyData_DAI32:  HALS_Object_GetPropertyData_DAI32: the object does not implement the property
coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_MIGServer.cpp:1806:HALS_Object_GetPropertyData_DAI32:  HALS_Object_GetPropertyData_DAI32: the object does not implement the property

coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_MIGServer.cpp:1806:HALS_Object_GetPropertyData_DAI32:  HALS_Object_GetPropertyData_DAI32: the object does not implement the property
coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_MIGServer.cpp:1806:HALS_Object_GetPropertyData_DAI32:  HALS_Object_GetPropertyData_DAI32: the object does not implement the property

ffmpeg: (CoreMediaIO) [com.apple.cmio:] CMIOHardware.cpp:339:CMIOObjectGetPropertyData the System is exiting
ffmpeg: (CoreMediaIO) [com.apple.cmio:] CMIOHardware.cpp:339:CMIOObjectGetPropertyData the System is exiting
ffmpeg: (CoreMediaIO) [com.apple.cmio:] CMIOHardware.cpp:339:CMIOObjectGetPropertyData the System is exiting
ffmpeg: (VDC) [com.apple.cmio:] CMIO_DALA_System.cpp:264:GetPropertyData error 1970171760 (<private>) getting property selector (<private>) scope (<private>) element 0
```

**Comment by devinroth** [comment]
I don’t see anything in that. What do you get when you use it with BlackHole the first to me when it actually works?

**Comment by fuerst** [comment]
That's a whole log from a first one:

```
Activity  ffmpeg: (libsystem_info.dylib) Retrieve User by ID
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOContextDescription.cpp:1128:PushStreamUsageToEngines_block_invoke:  IOContext 168: pushing 1 output stream(s) to 'AppleHDAEngineOutput:1B,0,1,1:0'
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOContextDescription.cpp:1132:PushStreamUsageToEngines_block_invoke:  IOContext 168: output stream[0] => Enabled: 'AppleHDAEngineOutput:1B,0,1,1:0'
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOContextDescription.cpp:1128:PushStreamUsageToEngines_block_invoke:  IOContext 169: pushing 1 input stream(s) to 'AppleHDAEngineInput:1B,0,1,0:1'
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOContextDescription.cpp:1132:PushStreamUsageToEngines_block_invoke:  IOContext 169: input stream[0] => Enabled: 'AppleHDAEngineInput:1B,0,1,0:1'
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOEngine2.cpp:1090:_UpdateTotalStreamUsage:  'BlackHole_UID': Updating 1 output stream(s)
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOEngine2.cpp:1090:_UpdateTotalStreamUsage:  'BlackHole_UID': Updating 1 input stream(s)
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOContextDescription.cpp:1128:PushStreamUsageToEngines_block_invoke:  IOContext 170: pushing 1 output stream(s) to 'BlackHole_UID'
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOContextDescription.cpp:1132:PushStreamUsageToEngines_block_invoke:  IOContext 170: output stream[0] => Enabled: 'BlackHole_UID'
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOContextDescription.cpp:1128:PushStreamUsageToEngines_block_invoke:  IOContext 170: pushing 1 input stream(s) to 'BlackHole_UID'
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOContextDescription.cpp:1132:PushStreamUsageToEngines_block_invoke:  IOContext 170: input stream[0] => Enabled: 'BlackHole_UID'
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOEngine2.cpp:1090:_UpdateTotalStreamUsage:  'BlackHole_UID': Updating 1 output stream(s)
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOEngine2.cpp:1090:_UpdateTotalStreamUsage:  'BlackHole_UID': Updating 1 input stream(s)
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOEngine2.cpp:1090:_UpdateTotalStreamUsage:  'com.rogueamoeba.InstantOn:0': Updating 0 output stream(s)
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOEngine2.cpp:1090:_UpdateTotalStreamUsage:  'com.rogueamoeba.InstantOn:0': Updating 0 input stream(s)
Activity  ffmpeg: (CoreFoundation) Loading Preferences From User CFPrefsD
Activity  ffmpeg: (TCC) TCCAccessRequest() IPC
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOContextDescription.cpp:1128:PushStreamUsageToEngines_block_invoke:  IOContext 170: pushing 1 output stream(s) to 'BlackHole_UID'
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOContextDescription.cpp:1132:PushStreamUsageToEngines_block_invoke:  IOContext 170: output stream[0] => Enabled: 'BlackHole_UID'
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOEngine2.cpp:1090:_UpdateTotalStreamUsage:  'BlackHole_UID': Updating 1 output stream(s)
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOContextDescription.cpp:1128:PushStreamUsageToEngines_block_invoke:  IOContext 170: pushing 1 input stream(s) to 'BlackHole_UID'
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOContextDescription.cpp:1132:PushStreamUsageToEngines_block_invoke:  IOContext 170: input stream[0] => Enabled: 'BlackHole_UID'
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOEngine2.cpp:1090:_UpdateTotalStreamUsage:  'BlackHole_UID': Updating 1 input stream(s)
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOContextDescription.cpp:1128:PushStreamUsageToEngines_block_invoke:  IOContext 170: pushing 1 input stream(s) to 'BlackHole_UID'
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOContextDescription.cpp:1132:PushStreamUsageToEngines_block_invoke:  IOContext 170: input stream[0] => Enabled: 'BlackHole_UID'
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOEngine2.cpp:1090:_UpdateTotalStreamUsage:  'BlackHole_UID': Updating 1 input stream(s)
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOContextDescription.cpp:1128:PushStreamUsageToEngines_block_invoke:  IOContext 170: pushing 1 output stream(s) to 'BlackHole_UID'
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOContextDescription.cpp:1132:PushStreamUsageToEngines_block_invoke:  IOContext 170: output stream[0] => Disabled: 'BlackHole_UID'
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOEngine2.cpp:1090:_UpdateTotalStreamUsage:  'BlackHole_UID': Updating 1 output stream(s)
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOContextDescription.cpp:1128:PushStreamUsageToEngines_block_invoke:  IOContext 170: pushing 1 input stream(s) to 'BlackHole_UID'
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOContextDescription.cpp:1132:PushStreamUsageToEngines_block_invoke:  IOContext 170: input stream[0] => Enabled: 'BlackHole_UID'
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOEngine2.cpp:1090:_UpdateTotalStreamUsage:  'BlackHole_UID': Updating 1 input stream(s)
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOContextDescription.cpp:1128:PushStreamUsageToEngines_block_invoke:  IOContext 170: pushing 1 output stream(s) to 'BlackHole_UID'
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOContextDescription.cpp:1132:PushStreamUsageToEngines_block_invoke:  IOContext 170: output stream[0] => Disabled: 'BlackHole_UID'
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOEngine2.cpp:1090:_UpdateTotalStreamUsage:  'BlackHole_UID': Updating 1 output stream(s)
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOContextDescription.cpp:1128:PushStreamUsageToEngines_block_invoke:  IOContext 170: pushing 1 input stream(s) to 'BlackHole_UID'
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOContextDescription.cpp:1132:PushStreamUsageToEngines_block_invoke:  IOContext 170: input stream[0] => Enabled: 'BlackHole_UID'
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOEngine2.cpp:1090:_UpdateTotalStreamUsage:  'BlackHole_UID': Updating 1 input stream(s)
Error     ffmpeg: (CoreMediaIO) [com.apple.cmio:] CMIOUnitFigBaseObjectImpl.c:246:CMIOUnitCreateFromDescription Invalid paramater
Error     ffmpeg: (CMIOBaseUnits) [com.apple.cmio:] CMIO_Unit_Input_HAL.cpp:985:GetPropertyInfo CMIOUInputBase::GetPropertyInfo failed for ID 102, Error: -67456
Activity  coreaudiod: (TCC) TCCAccessRequest() IPC
Error     ffmpeg: (CoreMediaIO) [com.apple.cmio:] CMIOHardware.cpp:339:CMIOObjectGetPropertyData the System is exiting
Error     ffmpeg: (CoreMediaIO) [com.apple.cmio:] CMIOHardware.cpp:339:CMIOObjectGetPropertyData the System is exiting
Error     ffmpeg: (CoreMediaIO) [com.apple.cmio:] CMIOHardware.cpp:339:CMIOObjectGetPropertyData the System is exiting
Error     ffmpeg: (VDC) [com.apple.cmio:] CMIO_DALA_System.cpp:264:GetPropertyData error 1970171760 (<private>) getting property selector (<private>) scope (<private>) element 0
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOEngine2.cpp:1090:_UpdateTotalStreamUsage:  'BlackHole_UID': Updating 1 output stream(s)
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOEngine2.cpp:1090:_UpdateTotalStreamUsage:  'BlackHole_UID': Updating 1 input stream(s)
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOEngine2.cpp:1090:_UpdateTotalStreamUsage:  'com.rogueamoeba.InstantOn:0': Updating 0 output stream(s)
Default   coreaudiod: [com.apple.coreaudio:AudioHAL_Server] HALS_IOEngine2.cpp:1090:_UpdateTotalStreamUsage:  'com.rogueamoeba.InstantOn:0': Updating 0 input stream(s)
```

**Comment by devinroth** [comment]
I think this is our error. 

Error     ffmpeg: (CoreMediaIO) [com.apple.cmio:] CMIOUnitFigBaseObjectImpl.c:246:CMIOUnitCreateFromDescription Invalid paramater

Error     ffmpeg: (CMIOBaseUnits) [com.apple.cmio:] CMIO_Unit_Input_HAL.cpp:985:GetPropertyInfo CMIOUInputBase::GetPropertyInfo failed for ID 102, Error: -67456


Not exactly sure what is causing this problem.

**Comment by fuerst** [comment]
Don't thinks so because this error occurs at the "good" call too. 

My guess is the "bad" call hangs somewhere in `get_audio_config()`, https://github.com/FFmpeg/FFmpeg/blob/master/libavdevice/avfoundation.m

Attached is the complete log of the good call (at 2020-01-09 23:17) and the bad call (at 2020-01-09 23:19).

[ffmpeg-coreaudiod.log](https://github.com/ExistentialAudio/BlackHole/files/4044629/ffmpeg-coreaudiod.log)

**Comment by devinroth** [comment]
My logic was that something is happening at the end of the first call that corrupts it in the next call. Regardless I'll take a deeper look hopefully this weekend.

**Comment by fuerst** [comment]
I see, that makes sense. This was at macOS 10.14 BTW.

**Comment by devinroth** [comment]
Couldn't reproduce the problem. Recording works every time. 

I did notice that ffmpeg won't record properly when overwriting a preexisting file. So first time it work fine and then 2nd time if I'm overwriting it doesn't. 

macOS 10.14.6
ffmpeg 4.2.2
BlackHole 0.2.5

**Comment by fuerst** [comment]
Really strange. Tried again using a fresh MacOS 10.14.6 (Virtualbox VM - see https://github.com/AlexanderWillner/runMacOSinVirtualBox). Installed BlackHole 0.2.5 and `brew install ffmpeg`. Running Safari and a Youtube video it worked multiple times. BlackHole is the only audio device in this setup. 

After enabling the Virtualbox audio device restarting the VM that changed - problem as described above. Removing the Virtualbox audio device did not help.

Since it works using Quicktime I guess I need to replace ffmpeg as audio recording tool.

**Comment by devinroth** [comment]
That's unfortunate. 

You should open a bug report with ffmpeg. 

Closing issue.

**Comment by fuerst** [comment]
I will. Thanks for your much appreciated help anyway!

**Comment by devinroth** [comment]
@fuerst try it again. My fix for #54 may fix your issues. I'm curious if it is somehow related.

**Comment by fuerst** [comment]
@devinroth Yes, Blackhole 0.2.6 fixed it - thanks a lot!
